bokehjs/src/lib/core/ui_events.ts

```diff
@@ -293,30 +301,141 @@ export class UIEvents implements EventListenerObject {
     return null
   }
 
-  protected _hit_test_frame(sx: number, sy: number): boolean {
-    return this.plot_view.frame.bbox.contains(sx, sy)
+  set_cursor(cursor: string = "default"): void {
+    this.hit_area.style.cursor = cursor
   }
 
-  protected _hit_test_canvas(sx: number, sy: number): boolean {
-    return this.plot_view.layout.bbox.contains(sx, sy)
+  protected _hit_test_frame(plot_view: PlotView, sx: number, sy: number): boolean {
+    return plot_view.frame.bbox.contains(sx, sy)
   }
 
+  protected _hit_test_canvas(plot_view: PlotView, sx: number, sy: number): boolean {
+    return plot_view.layout.bbox.contains(sx, sy)
+  }
+
+  protected _hit_test_plot(sx: number, sy: number): PlotView | null {
+    // TODO: z-index
+    for (const plot_view of this.canvas_view.plot_views) {
+      if (plot_view.layout.bbox.relative()/*XXX*/.contains(sx, sy))
+        return plot_view
+    }
+
+    return null
+  }
+
+  protected _prev_move: {sx: number, sy: number, plot_view: PlotView | null} | null = null
+
+  protected _curr_pan: {plot_view: PlotView} | null = null
+  protected _curr_pinch: {plot_view: PlotView} | null = null
+  protected _curr_rotate: {plot_view: PlotView} | null = null
+
   _trigger<E extends UIEvent>(signal: UISignal<E>, e: E, srcEvent: Event): void {
-    const gestures = this.toolbar.gestures
+    const {sx, sy} = e
+    const plot_view = this._hit_test_plot(sx, sy)
+    const curr_view = plot_view
+
+    const relativize_event = (_plot_view: PlotView): E => {
+      const [rel_sx, rel_sy] = [sx, sy] // plot_view.layout.bbox.relativize(sx, sy)
+      return {...e, sx: rel_sx, sy: rel_sy}
+    }
+
+    if (e.type == "panstart" || e.type == "pan" || e.type == "panend") {
```

Comment:
suggestion: this test could be `e.type.startstwith("pan")` (and similar for the other events) which reads a bit more immediately to me. Also what about moving all the conditional blocks to their own functions to shorten this function a bit?

Reply:
Those are not equivalent. It's written this way, because it allows the type system to narrow the type of `e` based on its `type` attribute. It's also better from forwards compatibility perspective.

Reply:
> Also what about moving all the conditional blocks to their own functions to shorten this function a bit?

In future PRs.